Updates the README and ensures correct jsdocs are present for all public APIs (#215)
ember-a11y/ember-a11y-testing: README.md

## Usage

> Setting up `ember-a11y-testing` in your tests can be done in one of two ways:

Author: Potato, PotAto :)

1. A one-time setup in your tests/test-helper.js file using `setupGlobalA11yHooks`
1. In individual tests using the `a11yAudit` test helper.

### aXe Options

When using the `a11yAudit` helper, you can pass in `axe-core` options.
These options are documented in the [axe-core API docs](https://www.deque.com/axe/axe-for-web/documentation/api-documentation/#options-parameter).
The rule definitions are documented on [dequeuniversity.com/rules](https://dequeuniversity.com/rules/axe/3.2).

Each of the following sections individually details how to set aXe options for your tests.

### `setupGlobalA11yHooks` Usage

> The `setupGlobalA11yHooks` function is intended to be imported and invoked a single time for your entire test suite.

Reviewer: Maybe mention here in the prose that it should probably be in `tests/test-helper.js`?

```ts
export function setupGlobalA11yHooks(

>   shouldAudit: InvocationStrategy,

Reviewer: I think referencing `InvocationStrategy` here actually makes things a bit confusing (since AFAICT you don't define what that actually is).

  audit: (...args: any[]) => PromiseLike<void> = a11yAudit
);
```

The `setupGlobalA11yHooks` function takes two parameters:

> - `shouldAudit`: A [predicate function](https://stackoverflow.com/a/1344021/769) that returns a `boolean` indicating whether or not to perform the audit.

Reviewer: The docs should indicate that it receives the helper name

Author: Updated to indicate this.

- `audit`: The audit function, which performs the `axe-core` audit, defaulting to `a11yAudit`. This allows you to potentially wrap the `a11yAudit` test helper with custom logic.

To use, import and invoke the global setup function, passing in your specific invocation strategy:

```js
import Application from 'my-app/app';
import config from 'my-app/config/environment';
import { setApplication } from '@ember/test-helpers';
import { start } from 'ember-qunit';
import { setupGlobalA11yHooks } from 'ember-a11y-testing/test-support';

> import { customInvocationStrategy } from './custom-invocation-strategy';
> 
> setApplication(Application.create(config.APP));
> 
> setupGlobalA11yHooks(customInvocationStrategy);

Reviewer: Maybe tweak this to be a bit simpler (and not reference the extra `custom-invocation-strategy` module):

```suggestion
setApplication(Application.create(config.APP));

setupGlobalA11yHooks(() => true));
```

start();
```

Examples of invocation strategies can be found in [this](https://github.com/ember-a11y/ember-a11y-testing/blob/50ef5f8fff4aa91d7a85b9feee5b1ce9bf380df9/tests/acceptance/setup-global-a11y-hooks-test.ts#L32) [repository's](https://github.com/ember-a11y/ember-a11y-testing/blob/50ef5f8fff4aa91d7a85b9feee5b1ce9bf380df9/tests/acceptance/setup-global-a11y-hooks-test.ts#L81) [tests](https://github.com/ember-a11y/ember-a11y-testing/blob/50ef5f8fff4aa91d7a85b9feee5b1ce9bf380df9/tests/acceptance/setup-global-a11y-hooks-test.ts#L135).

> #### Setting Options with `setupGlobalA11yHooks`

Reviewer: FWIW, I think that the `setRunOptions` API has nothing to do with `setupGlobalA11yHooks` specfiically. Any invocation of `a11yAudit` will use these options, not just the ones done by `setupGlobalA11yHooks`...

Author: I think it relates more to the global hooks, but you're correct in that you _can_ use the APIs anywhere. You can still pass the `options` into the `a11yAudit` calls directly, which is what I was trying to disambiguate.

        'color-contrast': {
          options: {
            noScroll: true,
          },
        },
      },
    });
  });

  // ...
});
```

Individual test level:

```javascript
import { module, test } from 'qunit';
import { a11yAudit } from 'ember-a11y-testing/test-support';
import { setRunOptions } from 'ember-a11y-testing';

module('some test module', function (hooks) {
  test('some test', function (assert) {
    setRunOptions({
      rules: {
        region: { enabled: true },
      },
      checks: {
        'color-contrast': {
          options: {
            noScroll: true,
          },
        },
      },
    });

    // ...
    a11yAudit();
    // ...
  });

  // ...
});
```

When using `setRunOptions` during a test, the options you set are automatically reset when the test completes.

### `a11yAudit` Usage

`ember-a11y-testing` also provides a test helper to run accessibility audits
on specific tests within your test suite.

_Note:_ any tests run with `ember-a11y-testing` will adjust the testing container
to occupy the entire screen. This is to simulate the actual application
environment, as browsers adjust styles at small sizes for accessibility reasons.

> It will reset itself at the conclusion of testing though.

Reviewer: Do we actually do this?

Reviewer 2: No...

Author: It was in the original docs. It struck me as weird, since I didn't see any code that would do this. I'll remove it.

#### Application Tests

For Application tests, the helper is an async test helper which can be used in a similar fashion to other `@ember/test-helpers` helpers:

```javascript
import { visit } from '@ember/test-helpers';
import { a11yAudit } from 'ember-a11y-testing/test-support';

module('Some module', function () {
  //...

  test('Some test case', async function (assert) {
    await visit('/');
    await a11yAudit();
    assert.ok(true, 'no a11y errors found!');
  });
});
```

> #### Integration and Unit Tests

Reviewer: Can we unify these sections (the ones for `Application Tests` and `Integration / Unit Tests`)? There really isn't anything unique to any of the types...

Author: Yep.

## Usage

Usage of `ember-a11y-testing` in your tests can be done in one of two ways:

1. A one-time setup in your tests/test-helper.js file using `setupGlobalA11yHooks`
1. In individual tests using the `a11yAudit` test helper.

> ### aXe Options

Reviewer 2: ```suggestion
### axe Options
```

When using the `a11yAudit` helper, you can pass in `axe-core` options.
These options are documented in the [axe-core API docs](https://www.deque.com/axe/axe-for-web/documentation/api-documentation/#options-parameter).

> The rule definitions are documented on [dequeuniversity.com/rules](https://dequeuniversity.com/rules/axe/3.2).

Reviewer 2: ```suggestion
The rule definitions are documented on [dequeuniversity.com/rules](https://dequeuniversity.com/rules/axe/4.0).
```

import { a11yAudit } from 'ember-a11y-testing/test-support';
import { setRunOptions } from 'ember-a11y-testing';

module('some test module', function (hooks) {
  test('some test', function (assert) {
    setRunOptions({
      rules: {
        region: { enabled: true },
      },
      checks: {
        'color-contrast': {
          options: {
            noScroll: true,
          },
        },
      },
    });

    // ...
    a11yAudit();
    // ...
  });

  // ...
});
```

When using `setRunOptions` during a test, the options you set are automatically reset when the test completes.

### `a11yAudit` Usage

`ember-a11y-testing` provides a test helper to run accessibility audits on specific tests within your test suite. The `a11yAudit` helper is an async test helper which can be used in a similar fashion to other `@ember/test-helpers` helpers:

In Application tests:

```javascript
import { visit } from '@ember/test-helpers';
import { a11yAudit } from 'ember-a11y-testing/test-support';

module('Some module', function () {
  //...

  test('Some test case', async function (assert) {
    await visit('/');
    await a11yAudit();
    assert.ok(true, 'no a11y errors found!');
  });
});
```

> The helper is also able to be used Integration/Unit tests like so:

Reviewer 2: ```suggestion
The helper is also able to be used in Integration/Unit tests like so:
```

  await visit('/');
  await a11yAudit('.modal', axeOptions);

>   assert.ok(true, 'no a11y errors found!');

Reviewer: Hehe, this really points out that the `a11yAudit` function throws 😭.

I very much want to change it to return something instead...

Author: Ya. I was thinking it would be nice to have an assertion API for when you don't have any other assertions in your test.

`assert.a11y().noViolationsExist();` and `assert.a11y().violationsExist();` or some such thing

});
```

### Optionally Running audits

> `ember-a11y-testing` also allows you to run audits only if `enableA11yAudit`

Reviewer: Hmm, I don't think this is quite true? We don't only run audits if the QP is set, we **always** run audits if the QP is set.

Author: Fixed up the verbiage.

}
```

### Logging violations to the console

> This addon provides the capability of summarizing all violations found during tests, and outputting those failures to the console once the test suite is completed. To enable this functionality, import `setupConsoleLogger` and invoke in your `tests/test-helper.js` file:

Reviewer: This says the name is `setupConsoleLogger` but the actual invocation below shows `setupConsoleReporter`, which is it?

Author: Oops. That was a mistake. Fixed.

## Future Plans

Now that your components and application tests can self-audit, the next step
going forward is to give helpful and meaningful feedback to developers. This
means easily highlighting areas with violations and giving suggestions on how to

> fix and improve them. Additionally, work will be done to tackle Ember-specific

Reviewer 2: Should probably remove this section as it talks about the visual audit component feature that was deprecated.

Author: Ya I was waiting for someone to notice this so I could delete it :)

> Now that your components and application tests can self-audit, the next step
> going forward is to give helpful and meaningful feedback to developers. This
> means easily highlighting areas with violations and giving suggestions on how to
> fix and improve them. Additionally, work will be done to tackle Ember-specific

Reviewer 2: ```suggestion
```
